File: R/tradingeconomics/R/federalReserve.R
source("R/functions.R")


#'Get Federal Reserve values from Trading Economics API
#'@export getFedReserve
#'
#'
#'@param category string.
#'@param indicator string.
#'@param outType string.
#''df' for data frame,
#''raw'(default) for list of unparsed data.
#'
#'@return Returns a list or data from Federal Reserve.
#'@section Notes:
#'Without credentials only sample information will be provided. Without a category, a list of the name of all states will be provided.
#'@seealso \code{\link{getCalendarData}}, \code{\link{getForecastData}}, \code{\link{getHistoricalData}} and \code{\link{getIndicatorData}}
#'@examples
#'\dontrun{ getFedReserve(indicator = 'arkansas', category = 'county')
#'getFedReserve(category = 'states')
#'}
#'

getFedReserve <- function(indicator = NULL, category = NULL, outType = NULL){
  base <- "https://api.tradingeconomics.com/fred"
  df_final = data.frame()

  if(!is.null(category == "name")) {
    url <- paste(base,"states", sep = '/')

  }
  if (category == "county"){
    url <- paste(base,"counties", paste(indicator, collapse = ','), sep = '/')

  }
  apikey_local <- .GlobalEnv$apiKey
  url <- paste(url, '?c=', apikey_local, sep = '')
  
  print(url)
  url <- URLencode(url)
  request <- GET(url)


  checkRequestStatus(http_status(request)$message)

  webResults <- do.call(rbind.data.frame, checkForNull(content(request)))

  df_final = rbind(df_final, webResults)
  Sys.sleep(0.5)


  if (is.null(outType)| identical(outType, 'lst')){
    df_final <- split(df_final , f = paste(df_final$county, df_final$name ))
  } else if (identical(outType, 'df')){
    df_final = df_final
  } else {
    stop('output_type options : df for data frame, lst(default) for list')
  }

  return(df_final)

}


#'Get Federal Reserve values from Trading Economics API
#'@export getFedReserveSnapshot
#'@param county string.
#' @param country string.
#' @param state string.
#' @param symbol string.
#' @param URL string.
#' @param outType string.
#''df' for data frame,
#''raw'(default) for list of unparsed data.
#'
#'@return Returns a list or data from Federal Reserve.
#'@section Notes:
#'Without credentials only sample information will be provided.
#'@seealso \code{\link{getCalendarData}}, \code{\link{getForecastData}}, \code{\link{getHistoricalData}} and \code{\link{getIndicatorData}}
#'@examples
#'\dontrun{ getFedReserveSnapshot('AGEXMAK2A647NCEN', outType = 'df')
#'getFedReserveSnapshot(county = 'arkansas')
#'getFedReserveSnapshot(county ='Pike County')
#'getFedReserveSnapshot(country = 'united states')
#'getFedReserveSnapshot(country = 'united states')
#'getFedReserveSnapshot(URL = '/united-states/income-inequality-in-aleutians-east
#'-borough-ak-fed-data.html' )
#'}
#'
getFedReserveSnapshot <- function(symbol = NULL, state = NULL, county = NULL, country = NULL, URL = NULL, outType = NULL){
  base <- "https://api.tradingeconomics.com/fred/snapshot"
  df_final = data.frame()
  apikey_local <- .GlobalEnv$apiKey
  if(!is.null(URL)){
    url <- paste(base, 'url', sep = '/')
    url <- paste(url, '?c=', apikey_local, sep = '')
    url <- paste(url, paste(URL, collapse = ','), sep = '&url=')

    url <- URLencode(url)
    request <- GET(url)
  }

  if(!is.null(symbol)){
    url <- paste(base, 'symbol', paste(symbol, collapse = ','), sep = '/')

  }
  if(!is.null(country)){
    url <- paste(base, 'country', paste(country, collapse = ','), sep = '/')
  }
  if(!is.null(state)){
    url <- paste(base, 'state', paste(state, collapse = ','), sep = '/')

  }
  if(!is.null(county)){
    url <- paste(base, 'county', paste(county, collapse = ','), sep = '/')

  }
  
  url <- paste(url, '?c=', apikey_local, sep = '')

  url <- URLencode(url)
  request <- GET(url)


  checkRequestStatus(http_status(request)$message)

  webResults <- do.call(rbind.data.frame, checkForNull(content(request)))

  df_final = rbind(df_final, webResults)
  Sys.sleep(0.5)


  if (is.null(outType)| identical(outType, 'lst')){
    df_final <- split(df_final , f = paste(df_final$Country, df_final$symbol, df_final$URL ))
  } else if (identical(outType, 'df')){
    df_final = df_final
  } else {
    stop('output_type options : df for data frame, lst(default) for list')
  }

  return(df_final)

}

#'Get Federal Reserve values from Trading Economics API
#'@export getFedHistorical
#'
#'@param symbol string.
#'@param initDate string with format: YYYY-MM-DD.
#'For example: '2011-01-01'.
#'@param endDate string with format: YYYY-MM-DD.
#'@param outType string.
#''df' for data frame,
#''raw'(default) for list of unparsed data.
#'
#'@return Returns a list or historical data from Federal Reserve.
#'@section Notes:
#'Without credentials only sample information will be provided. A symbol must be provided.
#'@seealso \code{\link{getCalendarData}}, \code{\link{getForecastData}}, \code{\link{getHistoricalData}} and \code{\link{getIndicatorData}}
#'@examples
#'\dontrun{ getFedHistorical('RACEDISPARITY005007', outType = 'df')
#'getFedHistorical(c('RACEDISPARITY005007','2020RATIO002013'), outType = 'df')
#'getFedHistorical(symbol='BAMLC0A1CAAAEY', initDate = '2018-05-01'  , outType = 'df')
#'getFedHistorical(symbol='BAMLC0A1CAAAEY', initDate = '2018-05-01', endDate = '2019-01-01'  , outType = 'df'))
#'}
#'

getFedHistorical <- function(symbol = NULL, initDate = NULL, endDate = NULL, outType = NULL){
  base <- "https://api.tradingeconomics.com/fred/historical"
  df_final = data.frame()
  apikey_local <- .GlobalEnv$apiKey
  if(is.null(symbol)) {
    print("A symbol is needed!")
  }else{
    url <- paste (base, paste(symbol, collapse = ','), sep = '/')
  }

  url <- paste(url, '?c=', apikey_local, sep = '')
  
  
  if (!is.null(initDate) & !is.null(endDate)){
    dateCheck(initDate)
    dateCheck(endDate)
    if (initDate > Sys.Date()) stop('Incorrect time period initDate!')
    if (initDate > endDate) stop('Incorrect time period initDate - endDate!')
    url <- paste(url, '&d1=' , paste(initDate, collapse = ','), sep = '')
    url <- paste(url, '&d2=', paste(endDate, collapse = ','), sep = '' )
  } 
  else if (!is.null(initDate)){
    dateCheck(initDate)
    if (initDate > Sys.Date()) stop('Incorrect time period initDate!')
    url <- paste(url, '&d1=' , paste(initDate, collapse = ','), sep = '' )
  }
 
  url <- URLencode(url)
  request <- GET(url)


  checkRequestStatus(http_status(request)$message)

  webResults <- do.call(rbind.data.frame, checkForNull(content(request)))

  df_final = rbind(df_final, webResults)
  Sys.sleep(0.5)


  if (is.null(outType)| identical(outType, 'lst')){
    df_final <- split(df_final , f = paste(df_final$symbol ))
  } else if (identical(outType, 'df')){
    df_final = df_final
  } else {
    stop('output_type options : df for data frame, lst(default) for list')
  }

  return(df_final)

}

File: nodejs/tradingeconomics/federalReserve.js
'use strict'

const auth = require('./auth.js');
const func = require('./functions.js');
const fetch = require('node-fetch');

//setting global variables to be used outside this module
global.counties = null;
global.country = null;
global.symbol = null;
global.county = null;
global.state = null;
global.URL = null;
global.historical_symbol = null;
global.start_date = null;
global.end_date = null;

//This function builds the path to get the API request:
/****************************************************************************************************  
   parameters:
    country, county, counties, states, symbol, URL, state, pike, historical(symbol), start_date, end_date

   example:
    getFred();
    getFred(counties = 'arkansas');
    getFred(country ='united states');      
    getFred(symbol ='ALLMARGATTN');              
    getFred(url ='/united-states/income-inequality-in-aleutians-east-borough-ak-fed-data.html' );              
    getFred(county ='pike' );              
    getFred(state ='tennessee' );                           
    getFred(historical_symbol ='RACEDISPARITY005007');
    getFred(historical_symbol = 'RACEDISPARITY005007', start_date='2017-05-01')
    getFred(historical_symbol = 'RACEDISPARITY005007', start_date='2017-05-01', end_date='2025-12-31')              

******************************************************************************************************/

function getFred(){

    try {
        var Data = '';
        var url = '';
    
        if ((counties, country, county, symbol, URL, state, historical_symbol) === null){    
            url =  '/fred/states'; 
        }
        if (counties != null){      
            url = '/fred/counties/' + counties;    
        }
        if (symbol != null){     
            url = '/fred/snapshot/symbol/' + symbol;    
        }
        if (URL != null){     
            url = '/fred/snapshot/url?';    
        }
        if (country != null){     
            url = '/fred/snapshot/country/' + country;    
        }
        if(state != null){
            url = '/fred/snapshot/state/' + state;
        }
        if(county != null){
            url = '/fred/snapshot/county/' + county;
        }
        if(historical_symbol != null){
            url = '/fred/historical/' + historical_symbol;
        }
        
        if(url.includes('?')){
            if(url.includes('url?')){
                Data = url_base + url + 'c='+ apikey.replace (' ','%20')+'&url=' + URL;
            }else{
            Data = url_base + url + '&c='+ apikey.replace (' ','%20');
        }
        }else{
            if (url.includes('/historical/')){
                if(start_date !== null && end_date == null){
                    Data = url_base + url + '?c='+ apikey.replace (' ','%20')+'&d1=' + start_date;
                }else if(start_date !== null && end_date !== null){
                    Data = url_base + url + '?c='+ apikey.replace (' ','%20')+'&d1=' + start_date + '&d2=' + end_date;
                }else{
                    Data = url_base + url + '?c='+ apikey.replace (' ','%20');
                }
        }else{
            Data = url_base + url + '?c='+ apikey.replace (' ','%20');
        }
        }
        return func.makeTheRequest(Data)
    
        // return fetch(Data)
        // .then(func.handleErrors)   
        // .then(function(response) {    
        //     return response.json(); // process it inside the `then` when calling the function       
        // }).catch(function (err) {
        //     return err.message;
        // });
    } catch (error) {
        throw error
    }
}

module.exports.getFred = getFred;









